src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/RazorLSPHandlerBase.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer;
+using Microsoft.AspNetCore.Razor.LanguageServer.Common;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Threading;
+using OmniSharpPosition = OmniSharp.Extensions.LanguageServer.Protocol.Models.Position;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp
+{
+    internal abstract class RazorLSPHandlerBase
+    {
+        private readonly int UndefinedDocumentVersion = -1;
+
+        public RazorLSPHandlerBase(
+            JoinableTaskContext joinableTaskContext,
+            ILanguageClientBroker languageClientBroker,
+            LSPDocumentManager documentManager,
+            LSPDocumentSynchronizer documentSynchronizer)
+        {
+            if (joinableTaskContext is null)
+            {
+                throw new ArgumentNullException(nameof(joinableTaskContext));
+            }
+
+            if (languageClientBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageClientBroker));
+            }
+
+            if (documentManager is null)
+            {
+                throw new ArgumentNullException(nameof(documentManager));
+            }
+
+            if (documentSynchronizer is null)
+            {
+                throw new ArgumentNullException(nameof(documentSynchronizer));
+            }
+
+            JoinableTaskFactory = joinableTaskContext.Factory;
+            LanguageClientBroker = languageClientBroker;
+            DocumentManager = documentManager;
+            DocumentSynchronizer = documentSynchronizer;
+        }
+
+        protected JoinableTaskFactory JoinableTaskFactory { get; }
+
+        protected ILanguageClientBroker LanguageClientBroker { get; }
+
+        protected LSPDocumentManager DocumentManager { get; }
+
+        protected LSPDocumentSynchronizer DocumentSynchronizer { get; }
+
+        protected virtual Task<TOut> RequestServerAsync<TIn, TOut>(
+            ILanguageClientBroker languageClientBroker,
+            string method,
+            LanguageServerKind serverKind,
+            TIn parameters,
+            CancellationToken cancellationToken)
+        {
+            if (languageClientBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageClientBroker));
+            }
+
+            if (string.IsNullOrEmpty(method))
+            {
+                throw new ArgumentException("message", nameof(method));
+            }
+
+            var contentType = RazorLSPContentTypeDefinition.Name;
```

Comment:
Interesting choice, so we could technically re-invoke some features on our own buffer. coolio